lib/datadog/grok/src/parse_grok_rules.rs

```diff
@@ -0,0 +1,378 @@
+                            }),
+                        }
+                    }
+                    Position::Only(pattern) => pattern.to_owned(),
+                })
+                .collect::<Vec<_>>(),
+            None => patterns.map(|p| p.to_owned()).collect::<Vec<_>>(),
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Converts each rule to a pure grok rule:
+///  - strips filters and "remembers" them to apply later
+///  - replaces references to previous rules with actual definitions
+///  - replaces match functions with corresponding regex groups
+fn purify_grok_pattern(
+    pattern: &GrokPattern,
+    mut filters: &mut HashMap<LookupBuf, Vec<GrokFilter>>,
+    prev_rule_filters: &mut HashMap<&str, HashMap<LookupBuf, Vec<GrokFilter>>>,
+    prev_rule_definitions: &mut HashMap<&str, String>,
+) -> Result<String, Error> {
+    let mut res = String::new();
+    if prev_rule_definitions.contains_key(pattern.match_fn.name.as_str()) {
+        // this is a reference to a previous rule - replace it and copy all destinations from the prev rule
+        res.push_str(
+            prev_rule_definitions
+                .get(pattern.match_fn.name.as_str())
+                .unwrap(),
+        );
+        if let Some(d) = prev_rule_filters.get(pattern.match_fn.name.as_str()) {
+            d.iter().for_each(|(path, function)| {
+                filters.insert(path.to_owned(), function.to_owned());
+            });
+        }
+    } else if pattern.match_fn.name == "regex"
+        || pattern.match_fn.name == "date"
+        || pattern.match_fn.name == "boolean"
+    {
+        // these patterns will be converted to named capture groups e.g. (?<http.status_code>[0-9]{3})
+        res.push_str("(?<");
+        if let Some(destination) = &pattern.destination {
+            res.push_str(destination.path.to_string().as_str());
+        }
+        res.push('>');
+        res.push_str(process_match_function(&mut filters, &pattern)?.as_str());
+        res.push(')');
+    } else {
+        // these will be converted to "pure" grok patterns %{PATTERN:DESTINATION} but without filters
+        res.push_str("%{");
+
+        res.push_str(process_match_function(&mut filters, &pattern)?.as_str());
+
+        if let Some(destination) = &pattern.destination {
+            write!(res, ":{}", destination.path).unwrap();
+        }
+        res.push('}');
+    }
+    Ok(res)
+}
+
+fn process_match_function(
+    filters: &mut HashMap<LookupBuf, Vec<GrokFilter>>,
+    pattern: &ast::GrokPattern,
+) -> Result<String, Error> {
+    let match_fn = &pattern.match_fn;
+    let result = match match_fn.name.as_ref() {
+        "regex" => {
+            if match_fn.args.is_some() {
+                if let ast::FunctionArgument::Arg(Value::Bytes(b)) =
+                    &match_fn.args.as_ref().unwrap()[0]
+                {
+                    return Ok(String::from_utf8_lossy(&b).to_string());
+                }
+            }
```

Comment:
I'd prefer the same pattern as above -- with a `Some(inner)` branch instead of matching on `.is_some()`

Reply:
Sure, done.